File: Codeforces/1431G.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <cmath>
#include <map>
#include <set>
#include <algorithm>
#include <queue>

#pragma GCC optimize("Ofast")
#pragma GCC target("avx,avx2,fma")
using namespace std;

template<typename T>
T readVal() {
    T n;
    cin >> n;
    return n;
}

template<typename T>
vector<T> readVector(T n) {
    vector<T> v(n);
    for (int i = 0; i < n; i++) {
        cin >> v[i];
    }
    return v;
}

template<typename T>
void printVector(vector<T> vec) {
    for (T x: vec) {
        cout << x << " ";
    }
    cout << endl;
}

const int INF = 1e9;

vector<vector<int>> dp1;
vector<vector<int>> dp2;
vector<int> arr;

int memoize(int pref, int K){
    if(K == 0){
        return (dp2[pref][K] = 0);
    }
    if(pref + 1 == K * 2){
        return (dp2[pref][K] = dp1[0][pref]);
    }
    if(pref + 1 < K * 2){
        return (dp2[pref][K] = -INF);
    }
    if(dp2[pref][K] != -1){
        return dp2[pref][K];
    }
    long long ans = memoize(pref - 1, K);
    //ans = max(ans, memoize(pref - 2, K - 1) + arr[pref] - arr[pref - 1]);
    for(int i = 0; i < pref; i ++){
        if(i % 2 == pref % 2) {
            long long val = dp1[i + 1][pref];
            long long len = (pref - (i + 1) + 1);
            if(len < 0 || K - len/2 < 0) continue;
            val += memoize(i, K - len / 2);
            ans = max(ans, val);
        }
    }
    return (dp2[pref][K] = ans);
}

void solve() {
    int N = readVal<int>();
    int K = readVal<int>();
    arr = readVector<int>(N);
    sort(arr.begin(), arr.end());
    dp1.resize(N);
    for (int i = 0; i < N; i++) {
        dp1[i].resize(N);
        for (int j = 0; j < N; j++) {
            dp1[i][j] = -1;
        }
    }
    for (int i = 0; i < N; i++) {
        for (int j = i + 1; j < N; j += 2) {
            int ans = 0;
            for(int k = i; k <= j; k++){
                if(k > i + j - k) {
                    ans += arr[k] - arr[i + j - k];
                }
            }
            dp1[i][j] = ans;
        }
    }
    dp2.resize(N);
    for(int i = 0; i < N; i++){
        dp2[i].resize(N);
        for(int j = 0; j < N; j++){
            dp2[i][j] = -1;
        }
    }
    for(int i = 0; i < N; i++){
        for(int j = 0; j <= K; j++){
            memoize(i, j);
        }
    }
    cout << dp2[N - 1][K] << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    solve();
}

File: Codeforces/1431G.kt
import java.util.*
import java.io.*


    fun main(args: Array<String>) {
        val sc = Scanner(System.`in`)
        val n = sc.nextInt()
        val k = sc.nextInt()
        val arr = IntArray(n)
        for (i in 0 until n) {
            arr[i] = sc.nextInt()
        }
        Arrays.sort(arr)
        val dp = Array(n) { IntArray(n) }
        for (i in n - 1 downTo 0) {
            for (j in i + 1 until n) {
                if (j % 2 == i % 2) {
                    dp[i][j] = 0
                } else {
                    if (j - i == 1) {
                        dp[i][j] = arr[j] - arr[i]
                    } else {
                        dp[i][j] = dp[i + 1][j - 1] + arr[j] - arr[i]
                    }
                }
            }
        }
        val ans = Array(n) { IntArray(n + 1) }
        for (i in 0 until n) {
            var j = 0
            while (j <= n) {
                if (j > i + 1 || j == 0) {
                    ans[i][j] = 0
                    j += 2
                    continue
                }
                if (i + 1 == j) {
                    ans[i][j] = dp[0][i]
                    j += 2
                    continue
                }
                ans[i][j] = ans[i - 1][j]
                for (l in 1..i) {
                    if (j - (i - l + 1) >= 0) {
                        ans[i][j] = Math.max(dp[l][i] + ans[l - 1][j - (i - l + 1)], ans[i][j])
                    }
                }
                j += 2
            }
        }
        println(ans[n - 1][2 * k])
    }
